session/pingpong/exchange_message_tracker.go

```diff
@@ -59,32 +117,135 @@ var ErrInvoiceMissmatch = errors.New("invoice mismatch")
 // Start starts the message exchange tracker. Blocks.
 func (emt *ExchangeMessageTracker) Start() error {
 	log.Debug("Starting...")
+	addr, err := crypto.GenerateChannelAddress(emt.identity.Address, emt.registryAddress, emt.channelImplementation)
+	if err != nil {
+		return errors.Wrap(err, "could not generate channel address")
+	}
+	emt.channelAddress = identity.FromAddress(addr)
+
+	emt.timeTracker.StartTracking()
 	for {
 		select {
 		case <-emt.stop:
 			return nil
 		case invoice := <-emt.invoiceChan:
 			log.Debug("Invoice received", invoice)
-			err := emt.issueExchangeMessage(invoice)
+			err := emt.isInvoiceOK(invoice)
+			if err != nil {
+				return errors.Wrap(err, "invoice not valid")
+			}
+
+			err = emt.issueExchangeMessage(invoice)
 			if err != nil {
 				return err
 			}
+
+			err = emt.consumerInvoiceStorage.Store(emt.peer, invoice)
+			if err != nil {
+				return errors.Wrap(err, "could not store invoice")
+			}
+
+		}
+	}
+}
+
+const grandTotalKey = "consumer_grand_total"
+
+func (emt *ExchangeMessageTracker) getGrandTotalPromised() (uint64, error) {
+	res, err := emt.consumerTotalsStorage.Get(grandTotalKey)
+	if err != nil {
+		if err == ErrNotFound {
+			log.Debug("no previous invoice grand total, assuming zero")
+			return 0, nil
+		}
+	}
+	return res, errors.Wrap(err, "could not get previous grand total")
```

Comment:
It would be more clear if you put this error inside if err != nil

```go
if err != nil {
   if err == ErrNotFound {
      log.Debug("no previous invoice grand total, assuming zero")
      return 0, nil
   }
   return 0, errors.Wrap(err, "could not get previous grand total")
}
return res, nil
```